test/f3_test.go

```diff
@@ -323,12 +324,17 @@ func (e *testEnv) waitForManifestChange(prev *manifest.Manifest, timeout time.Du
 	}, timeout, ManifestSenderTimeout)
 }
 
-func newTestEnvironment(t *testing.T, n int, dynamicManifest bool) testEnv {
+func newTestEnvironment(t *testing.T, n int, dynamicManifest bool) *testEnv {
 	ctx, cancel := context.WithCancel(context.Background())
 	grp, ctx := errgroup.WithContext(ctx)
-	env := testEnv{t: t, errgrp: grp, testCtx: ctx, net: mocknet.New()}
+	env := &testEnv{t: t, errgrp: grp, testCtx: ctx, net: mocknet.New()}
+
+	// Cleanup on exit.
 	env.t.Cleanup(func() {
 		cancel()
+		for _, n := range env.nodes {
+			require.NoError(env.t, n.f3.Stop(context.Background()))
+		}
 		require.NoError(env.t, env.errgrp.Wait())
 	})
```

Comment:
This wasn't needed by the final version of this patch, but it was for some earlier iterations and, in general, is less likely to deadlock on exit. Basically, it ensures that we stop all F3 nodes:

1. Only once.
2. Before waiting.